beartype/typing.py

```diff
@@ -288,3 +288,211 @@
+                bases_pass_muster = True
+
+                for base in cls.__bases__:
+                    if base is cls or base.__name__ in (
+                        "Protocol",
+                        "Generic",
+                        "object",
+                    ):
+                        continue
+
+                    if not isinstance(inst, base):
+                        bases_pass_muster = False
+                        break
+
+                cls._abc_inst_check_cache[
+                    inst_t
+                ] = bases_pass_muster and cls._check_only_my_attrs(inst)
+
+                return cls._abc_inst_check_cache[inst_t]
+
+        def _check_only_my_attrs(cls, inst: Any) -> bool:
+            attrs = set(cls.__dict__)
+            attrs.update(cls.__dict__.get("__annotations__", {}))
+            attrs.intersection_update(_get_protocol_attrs(cls))
+
+            for attr in attrs:
+                if not hasattr(inst, attr):
+                    return False
+                elif callable(getattr(cls, attr, None)) and getattr(inst, attr) is None:
+                    return False
+
+            return True
+
+    @runtime_checkable
+    class Protocol(_Protocol, Generic, metaclass=_CachingProtocolMeta):  # type: ignore [no-redef]
+        """
+        ``@beartype``-compatible (i.e., ``@runtime_checkable``) drop-in
+        replacement for :class:`typing.Protocol` that can lead to significant
+        performance improvements. Uses :class:`_CachingProtocolMeta` to cache
+        :func:`isinstance` check results.
+
+        .. code-block:: python
+          :linenos:
+
+          >>> from abc import abstractmethod
+          >>> from beartype import beartype
+          >>> from beartype.typing import Protocol
+
+          >>> class MyBearProtocol(Protocol):  # <-- runtime-checkable through inheritance
+          ...   @abstractmethod
+          ...   def myfunc(self, arg: int) -> str:
+          ...     pass
+
+          >>> my_thing: MyBearProtocol = MyImplementation()
+          >>> isinstance(my_thing, MyBearProtocol)
+          True
+
+          >>> @beartype
+          ... def do_somthing(thing: MyBearProtocol) -> None:
+          ...   thing.myfunc(0)
+        """
+        __slots__: Union[str, Iterable[str]] = ()
+
+    from typing import SupportsAbs as _SupportsAbs
+    from typing import SupportsBytes as _SupportsBytes
+    from typing import SupportsComplex as _SupportsComplex
+    from typing import SupportsFloat as _SupportsFloat
+    from typing import SupportsIndex as _SupportsIndex
+    from typing import SupportsInt as _SupportsInt
+    from typing import SupportsRound as _SupportsRound
+
+    class SupportsAbs(_SupportsAbs[_T_co], Protocol, Generic[_T_co]):  # type: ignore [no-redef]
+        "A caching version of :class:`typing.SupportsAbs`."
+        __slots__: Union[str, Iterable[str]] = ()
+
+    class SupportsBytes(_SupportsAbs, Protocol):  # type: ignore [no-redef]
```

Comment:
Whoops! This should be `class SupportsBytes(_SupportsBytes, Protocol): …`. I'll fix this….

Reply:
Corrected as of a35e6de9edce0cdbfc76fe047bd9823ec0d1bfd8.